src/layers/mpls.rs

```diff
@@ -0,0 +1,72 @@
+//! Multi Label Protocol Switching
+
+use std::convert::TryInto;
+
+use serde::Serialize;
+
+use crate::errors::Error;
+use crate::layers::ethernet;
+use crate::types::{ETHERTYPE_MPLS_MULTICAST, ETHERTYPE_MPLS_UNICAST};
+use crate::Layer;
+
+use super::ipv4::IPv4;
+
+/// Default Header Length for MPLS Packets
+pub const MPLS_HDR_LEN: usize = 4_usize;
+
+// Register Ourselves to the Ethernet layer, as this is a 2.5 layer protocol
+pub(crate) fn register_defaults() -> Result<(), Error> {
+    ethernet::register_ethertype(ETHERTYPE_MPLS_UNICAST, MPLS::creator)?;
+    ethernet::register_ethertype(ETHERTYPE_MPLS_MULTICAST, MPLS::creator)
+}
+
+#[derive(Debug, Default, Serialize)]
+pub struct MPLS {
+    //Make built-in types
+    #[serde(serialize_with = "crate::types::hex::serialize_lower_hex_u32")]
+    label: u32, //This is only 20 bits
+    #[serde(serialize_with = "crate::types::hex::serialize_lower_hex_u8")]
+    exp: u8, //This is only 3 bits
+    #[serde(serialize_with = "crate::types::hex::serialize_lower_hex_u8")]
+    bos: u8, //This is only 1 bit
+    #[serde(serialize_with = "crate::types::hex::serialize_lower_hex_u8")]
+    ttl: u8, //This is 1 byte
+}
+
+impl MPLS {
+    pub fn creator() -> Box<dyn Layer + Send> {
+        Box::<MPLS>::default()
+    }
+}
+
+impl Layer for MPLS {
+    fn decode_bytes(
+        &mut self,
+        bytes: &[u8],
+    ) -> Result<(Option<Box<dyn Layer + Send>>, usize), Error> {
+        if bytes.len() < MPLS_HDR_LEN {
```

Comment:
You might have to do all of this in a `loop`. So something like 

```rust 
loop { 
    // Check for length of the remaining buffer if less than 4 bytes error out.

    // Dissect the label

    // If `bos` break
    
    // update the decoded length

} 

  // Return `Ok` with `IPv4::creator` and total decoded length. 
```

Reply:
I have implemented the looping feature, however, I can't find an appropriate packet for testing the payload following the multiple label stack(though the looping for the stack itself is working). This is because, the ones that use MPLS with multiple labels follow a protocol different than IPv4. For example, in this [packet](https://packetlife.net/media/captures/EoMPLS.cap), the MPLS label stack is followed by an Ethernet Control Word, IEEE 802.3 Ethernet, Logical-Link Control and the Spanning-Tree Protocol.